### DISCUSSION TITLE
Should `CURLU_GUESS_SCHEME` always guess `http://` if the host name doesn't have a 3rd level domain?

### DISCUSSION THREAD

**Discussion by emanuele6** [curl/curl#11161] [Ideas]
`libcurl`'s `curl_set_url()` function when used with the `CURLU_GUESS_SCHEME` flag, and, by consequence, the `curl` tool when it interprets its URL arguments, will "guess" the scheme for URLs that do not have an explict scheme based on their "outermost sub-domain name", or "http://" otherwise, e.g.

* `imap.gmail.com`
  => `imap://imap.google.com/`
* `emanuele6:hunter2@smtp.example.org`
  => `smtp://emanuele6:hunter2@smtp.example.org/`
* `ftp.gnu.org/gnu/wget/`
  => `ftp://ftp.gnu.org/gnu/wget/`
* `curl.se`
  => `http://curl.se/`

This generally makes sense, but I noticed that `CURLU_GUESS_SCHEME` will also guess based on the name of the 2nd level domain: e.g.

* `imap.com`: the website of "International Mergers & Acquisitions Partners" that does not host an IMAP server
  => `imap://imap.com/`
  ```console
  $ time curl imap.com
  curl: (7) Failed to connect to imap.com port 143 after 25 ms: Couldn't connect to server
  
  real    0m0.034s
  user    0m0.000s
  sys     0m0.011s
  ```
* `dict.com`: A dictionary/translation website... that does not host a DICT server
  => `dict://dict.com/`
  ```console
  $ time curl dict.com
  curl: (28) Failed to connect to dict.com port 2628 after 129286 ms: Couldn't connect to server
  
  real    2m9.298s
  user    0m0.010s
  sys     0m0.012s
  ```
* `ldap.com`: A website about the LDAP protocol... that does not host a LDAP server
  => `ldap://ldap.com/`
  ```console
  $ time src/curl ldap.com # my distro's curl didn't have LDAP support
  curl: (28) Failed to connect to ldap.com port 389 after 258699 ms: Couldn't connect to server
  
  real    4m18.731s
  user    0m0.030s
  sys     0m0.026s
  ```

I think guessing the scheme based on the second level domain doesn't make much sense. In my opinion, it would make more sense to always guess "http://" if the outermost domain name is not 3rd level or higher.

The only domain I can think of for which guessing on the second level domain makes sense is [dict.org](http://dict.org), and probably a bunch of users are accessing it with `curl` without specifying the `dict://` scheme, so it may be worth adding a special rule that guesses `dict://` for specifically "dict.org" (or "dict.tld" in general) if we decide to change the behaviour of `CURLU_GUESS_SCHEME`.

What do you think?

&nbsp;emanuele6

**Comment by bagder** [comment]
The protocol guessing thing is weak and generally not very helpful to begin with. But yes, the guess should probably be different/disabled without a 3rd level domain. But is it worth poking at?
